src/Common/PathUtilities.cs

```diff
@@ -5,41 +5,27 @@ namespace Microsoft.DotNet;
 
 static class PathUtilities
 {
-    const int S_IRUSR = 256;
-    const int S_IWUSR = 128;
-    const int S_IXUSR = 64;
-    const int S_IRWXU = S_IRUSR | S_IWUSR | S_IXUSR; // 700 (octal) Permissions 
-
-    const int MAX_NUM_DIRECTORY_CREATE_RETRIES = 2;
-
     public static string CreateTempSubdirectory()
-    {
-        return CreateTempSubdirectoryRetry(0);
-    }
-
-    [DllImport("libc", SetLastError = true)]
-    private static extern int mkdir(string pathname, int mode);
-    private static string CreateTempSubdirectoryRetry(int attemptNo)
     {
         string path = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
-        if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
-        {
-            int mkdirStatusCode = mkdir(path, S_IRWXU);
-            if (mkdirStatusCode != 0)
-            {
-                int errno = Marshal.GetLastWin32Error();
-                if (Directory.Exists(path) && attemptNo < MAX_NUM_DIRECTORY_CREATE_RETRIES)
-                {
-                    return CreateTempSubdirectoryRetry(attemptNo + 1);
-                }
-                else
-                    throw new IOException($"Failed to create a temporary subdirectory {path} with mkdir, error code: {errno}");
-            }
-        }
+
+#if NETFRAMEWORK
+        Directory.CreateDirectory(path);
+#else
+        if (OperatingSystem.IsWindows())
+            Directory.CreateDirectory(path);
         else
         {
-            Directory.CreateDirectory(path);
+            UnixFileMode desiredMode = UnixFileMode.UserRead | UnixFileMode.UserWrite | UnixFileMode.UserExecute;
+
+            do 
+            {
+                path = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
+                Directory.CreateDirectory(path, desiredMode);
+            } while (File.GetUnixFileMode(path) != desiredMode);
```

Comment:
Should we do this forever? It might be good to stop after a certain point, I don't think we want to hang on this forever and it would be better to give a clear error.

Reply:
Fair point

Reply:
Ok, I think this is OK, but I'm going to go through some channels to see if we can get additional verification on this.